flashlight/autograd/backend/cpu/Conv2D.cpp

```diff
@@ -42,9 +42,26 @@ Variable conv2d(
     int py,
     int dx,
     int dy,
-    int groups) {
+    int groups,
+    bool useMixedPrecision,
+    const ConvPerfAuditPtr audit) {
```

Comment:
Is the `ConvPerfAuditPtr` also for debugging benchmarking? I'd prefer if we removed it before merging.

Reply:
`ConvPerfAuditPtr` is not for debugging. It helps with choosing the convolution algorithm that yields the highest performance. Depending on the convolution kernel, there are cases that performing the convolution in F32 can actually yield a better performance (since not all algorithms can support half precision). In those cases, even if the data is in F16, it is beneficial to perform the convolution in F32. I understand that it adds complication, so we can remove it. However, if we remove it, there will be significant performance loss on some networks.

Reply:
Got it — that's helpful. Didn't quite understand what was happening there.

My primary question is — why do we actually need this? Why is the choice of algorithm empirical like this — can't cuDNN select this ahead of time based on heuristics similar to these and the cudnn math mode?

We can definitely have some sort of stateful thing which keeps tallies of performance for the operator — need to think about the implementation a bit.

Reply:
Also want to add — because the shape of the convolutions that we perform changes so frequently, we never use `cuDNN` benchmark mode, so those heuristics don't really do much for us. I'd be interested in seeing a benchmark with and without the Conv audit.